sound/soc/sof/sof-client.c

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
+//
+
+#include <linux/device.h>
+#include <linux/list.h>
+#include <linux/mutex.h>
+#include <linux/virtual_bus.h>
+#include "sof-client.h"
+
+static DEFINE_MUTEX(sof_client_mutex);
+
+static void sof_virtdev_release(struct virtbus_device *vdev)
+{}
+
+int sof_client_dev_register(struct snd_sof_dev *sdev,
+			    const char *name)
+{
+	struct sof_client_dev *cdev;
+	struct virtbus_device *vdev;
+	int ret;
+
+	cdev = devm_kzalloc(sdev->dev, sizeof(*cdev), GFP_KERNEL);
+	if (!cdev)
+		return -ENOMEM;
+
+	vdev = &cdev->vdev;
+	vdev->name = name;
+	vdev->dev.parent = sdev->dev;
+	vdev->release = sof_virtdev_release;
+
+	/* register virtbus device for the client */
+	ret = virtbus_dev_register(&cdev->vdev);
+	if (ret < 0)
+		return ret;
+
+	/* add to list of SOF client devices */
+	mutex_lock(&sof_client_mutex);
```

Comment:
if you march through the sdev->client_list in the core (for e.g. PM stuff), then you probably want the core to take the mutex as well?
in that case, shouldn't the mutex part of sdev ?

Reply:
@plbossart I started off with the client_mutex being part of sdev but changed it because of the doubt that sof_client_dev_unregister() could get called when the client module is removed maybe through an rmmod. Not sure if this is even possible though. If the client_dev_unregister() would always be invoked by the core, then I suppose I can move the mutex to be part of snd_sof_dev as well

Reply:
@plbossart I'm still torn about this. Would there ever be a case where a client could get unregistered other than by the core?

Reply:
not sure. I think it's possible that the driver is unbound (and the module removed), but removing a device is typically done by the entity creating it

Reply:
ok I moved the mutex to be part of snd_sof_dev too.